jax/_src/random.py

```diff
@@ -958,7 +958,13 @@ def _bernoulli(key, p, shape) -> Array:
   else:
     _check_shape("bernoulli", shape, np.shape(p))
 
-  return uniform(key, shape, lax.dtype(p)) < p
+  # we could return uniform(key, shape) < p. But uniform sacrifices some
+  # resolution, so instead we sample in the space of the generated bits.
+  nbits = dtypes.finfo(p.dtype).bits
+  unsigned_dtype = UINT_DTYPES[nbits]
+  samples = bits(key, shape, unsigned_dtype)
+  cutoff = (p * _lax_const(p, 1 << nbits)).astype(unsigned_dtype)
+  return (p >= 1) | (samples < cutoff)
```

Comment:
Do you also need to protect against negative p? It might wrap around and give a wrong answer.

```
>>> np.float32(-0.5 * 2**32).astype(np.uint32)
2147483648
```

Reply:
Done